File: app/src/main/java/forpdateam/ru/forpda/fragments/devdb/DeviceFragment.java
package forpdateam.ru.forpda.fragments.devdb;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.util.Pair;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewStub;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.nostra13.universalimageloader.core.ImageLoader;
import com.nostra13.universalimageloader.core.assist.FailReason;
import com.nostra13.universalimageloader.core.listener.SimpleImageLoadingListener;
import com.robohorse.pagerbullet.PagerBullet;

import java.util.ArrayList;
import java.util.Observer;

import forpdateam.ru.forpda.App;
import forpdateam.ru.forpda.R;
import forpdateam.ru.forpda.api.devdb.models.Device;
import forpdateam.ru.forpda.fragments.TabFragment;
import forpdateam.ru.forpda.fragments.devdb.device.comments.CommentsFragment;
import forpdateam.ru.forpda.fragments.devdb.device.posts.PostsFragment;
import forpdateam.ru.forpda.fragments.devdb.device.specs.SpecsFragment;
import forpdateam.ru.forpda.fragments.notes.NotesAddPopup;
import forpdateam.ru.forpda.imageviewer.ImageViewerActivity;
import forpdateam.ru.forpda.rxapi.RxApi;
import forpdateam.ru.forpda.utils.IntentHandler;
import forpdateam.ru.forpda.utils.Utils;
import forpdateam.ru.forpda.utils.rx.Subscriber;

/**
 * Created by radiationx on 08.08.17.
 */

public class DeviceFragment extends TabFragment {
    public final static String ARG_DEVICE_ID = "DEVICE_ID";
    private String deviceId = "";
    private Subscriber<Device> mainSubscriber = new Subscriber<>(this);
    private PagerBullet imagesPager;
    private TabLayout tabLayout;
    private TextView rating;
    private ViewPager fragmentsPager;
    private ProgressBar progressBar;
    private Device currentData;
    private RelativeLayout toolbarContent;
    private Observer statusBarSizeObserver = (observable1, o) -> {
        if (toolbarContent != null) {
            CollapsingToolbarLayout.LayoutParams params = (CollapsingToolbarLayout.LayoutParams) toolbarContent.getLayoutParams();
            params.topMargin = /*App.getToolBarHeight(toolbarLayout.getContext()) +*/ App.getStatusBarHeight();
            toolbarContent.setLayoutParams(params);
        }
    };

    public DeviceFragment() {
        configuration.setDefaultTitle(App.getInstance().getString(R.string.fragment_title_device));
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            deviceId = getArguments().getString(ARG_DEVICE_ID, deviceId);
        }

        if (getChildFragmentManager().getFragments() != null) {
            FragmentTransaction transaction = getChildFragmentManager().beginTransaction();
            for (Fragment fragment : getChildFragmentManager().getFragments()) {
                transaction.remove(fragment);
            }
            transaction.commit();
            getChildFragmentManager().executePendingTransactions();
        }

    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
    }


    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        setCardsBackground();
        baseInflateFragment(inflater, R.layout.fragment_device);
        ViewStub viewStub = (ViewStub) findViewById(R.id.toolbar_content);
        viewStub.setLayoutResource(R.layout.toolbar_device);
        toolbarContent = (RelativeLayout) viewStub.inflate();

        imagesPager = (PagerBullet) findViewById(R.id.images_pager);
        progressBar = (ProgressBar) findViewById(R.id.progress_bar);
        rating = (TextView) findViewById(R.id.item_rating);
        viewsReady();

        toolbarTitleView.setShadowLayer(App.px2, 0, 0, App.getColorFromAttr(getContext(), R.attr.colorPrimary));
        toolbarSubtitleView.setShadowLayer(App.px2, 0, 0, App.getColorFromAttr(getContext(), R.attr.colorPrimary));

        toolbarLayout.setExpandedTitleColor(Color.TRANSPARENT);
        toolbarLayout.setCollapsedTitleTextColor(Color.TRANSPARENT);
        toolbarLayout.setTitleEnabled(false);
        AppBarLayout.LayoutParams params = (AppBarLayout.LayoutParams) toolbarLayout.getLayoutParams();
        params.setScrollFlags(AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL | AppBarLayout.LayoutParams.SCROLL_FLAG_EXIT_UNTIL_COLLAPSED); // list other flags here by |
        toolbarLayout.setLayoutParams(params);


        CollapsingToolbarLayout.LayoutParams newParams = (CollapsingToolbarLayout.LayoutParams) toolbar.getLayoutParams();
        newParams.setCollapseMode(CollapsingToolbarLayout.LayoutParams.COLLAPSE_MODE_PIN);
        newParams.bottomMargin = App.px48;
        toolbar.setLayoutParams(newParams);
        toolbar.requestLayout();


        tabLayout = new TabLayout(getContext());
        CollapsingToolbarLayout.LayoutParams tabparams = new CollapsingToolbarLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT, Gravity.BOTTOM);
        tabparams.setCollapseMode(CollapsingToolbarLayout.LayoutParams.COLLAPSE_MODE_PIN);
        tabLayout.setLayoutParams(tabparams);
        tabLayout.setTabMode(TabLayout.MODE_SCROLLABLE);
        toolbarLayout.addView(tabLayout);


        fragmentsPager = (ViewPager) findViewById(R.id.view_pager);
        tabLayout.setupWithViewPager(fragmentsPager);

        imagesPager.setIndicatorTintColorScheme(App.getColorFromAttr(getContext(), R.attr.default_text_color), App.getColorFromAttr(getContext(), R.attr.second_text_color));

        App.getInstance().addStatusBarSizeObserver(statusBarSizeObserver);
        return view;
    }

    private MenuItem copyLinkMenuItem, shareMenuItem, noteMenuItem, toBrandMenuItem, toBrandsMenuItem;

    @Override
    protected void addBaseToolbarMenu() {
        super.addBaseToolbarMenu();
        copyLinkMenuItem = getMenu().add(R.string.copy_link)
                .setOnMenuItemClickListener(item -> {
                    Utils.copyToClipBoard("https://4pda.ru/devdb/" + currentData.getId());
                    return true;
                });

        shareMenuItem = getMenu().add(R.string.share)
                .setOnMenuItemClickListener(item -> {
                    Utils.shareText("https://4pda.ru/devdb/" + currentData.getId());
                    return true;
                });

        noteMenuItem = getMenu().add(R.string.create_note)
                .setOnMenuItemClickListener(item -> {
                    String title = "DevDb: " + currentData.getBrandTitle() + " " + currentData.getTitle();
                    String url = "https://4pda.ru/devdb/" + currentData.getId();
                    NotesAddPopup.showAddNoteDialog(getContext(), title, url);
                    return true;
                });

        toBrandMenuItem = getMenu().add(R.string.devices)
                .setOnMenuItemClickListener(item -> {
                    IntentHandler.handle("https://4pda.ru/devdb/" + currentData.getCatId() + "/" + currentData.getBrandId());
                    return true;
                });

        toBrandsMenuItem = getMenu().add(R.string.devices)
                .setOnMenuItemClickListener(item -> {
                    IntentHandler.handle("https://4pda.ru/devdb/" + currentData.getCatId());
                    return true;
                });

        refreshToolbarMenuItems(false);
    }

    @Override
    protected void refreshToolbarMenuItems(boolean enable) {
        super.refreshToolbarMenuItems(enable);
        if (enable) {
            copyLinkMenuItem.setEnabled(true);
            shareMenuItem.setEnabled(true);
            noteMenuItem.setEnabled(true);
            toBrandMenuItem.setVisible(true);
            toBrandsMenuItem.setVisible(true);
        } else {
            copyLinkMenuItem.setEnabled(false);
            shareMenuItem.setEnabled(false);
            noteMenuItem.setEnabled(false);
            toBrandMenuItem.setVisible(false);
            toBrandsMenuItem.setVisible(false);
        }
    }

    @Override
    public void loadData() {
        super.loadData();
        refreshToolbarMenuItems(false);
        progressBar.setVisibility(View.VISIBLE);
        mainSubscriber.subscribe(RxApi.DevDb().getDevice(deviceId), this::onLoad, new Device());
    }

    private void onLoad(Device device) {
        currentData = device;
        progressBar.setVisibility(View.GONE);
        toBrandMenuItem.setTitle(currentData.getCatTitle() + " " + currentData.getBrandTitle());
        toBrandsMenuItem.setTitle(currentData.getCatTitle());
        refreshToolbarMenuItems(true);
        setTitle(currentData.getTitle());
        setSubtitle(currentData.getCatTitle() + " > " + currentData.getBrandTitle());


        ArrayList<String> urls = new ArrayList<>();
        ArrayList<String> fullUrls = new ArrayList<>();
        for (Pair<String, String> pair : currentData.getImages()) {
            urls.add(pair.first);
            fullUrls.add(pair.second);
        }
        ImagesAdapter imagesAdapter = new ImagesAdapter(getContext(), urls, fullUrls);
        imagesPager.setAdapter(imagesAdapter);

        FragmentPagerAdapter pagerAdapter = new FragmentPagerAdapter(getChildFragmentManager(), currentData);
        fragmentsPager.setAdapter(pagerAdapter);

        if (device.getRating() > 0) {
            rating.setText(Integer.toString(device.getRating()));
            rating.setBackground(App.getDrawableAttr(rating.getContext(), R.attr.count_background));
            rating.getBackground().setColorFilter(RxApi.DevDb().getColorFilter(device.getRating()));
            rating.setVisibility(View.VISIBLE);
            if (device.getComments().size() > 0) {
                rating.setClickable(true);
                rating.setOnClickListener(v -> fragmentsPager.setCurrentItem(1, true));
            }

        } else {
            rating.setVisibility(View.GONE);
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        App.getInstance().removeStatusBarSizeObserver(statusBarSizeObserver);
    }

    private class FragmentPagerAdapter extends android.support.v4.app.FragmentPagerAdapter {
        private Device device;
        private ArrayList<Fragment> fragments = new ArrayList<>();
        private ArrayList<String> titles = new ArrayList<>();

        public FragmentPagerAdapter(FragmentManager fm, Device device) {
            super(fm);
            this.device = device;
            if (this.device.getSpecs().size() > 0) {
                fragments.add(new SpecsFragment().setDevice(this.device));
                titles.add(App.getInstance().getString(R.string.device_page_specs));
            }
            if (this.device.getComments().size() > 0) {
                fragments.add(new CommentsFragment().setDevice(this.device));
                titles.add(App.getInstance().getString(R.string.device_page_comments) + " (" + this.device.getComments().size() + ")");
            }
            if (this.device.getDiscussions().size() > 0) {
                fragments.add(new PostsFragment().setSource(PostsFragment.SRC_DISCUSSIONS).setDevice(this.device));
                titles.add(App.getInstance().getString(R.string.device_page_discussions) + " (" + this.device.getDiscussions().size() + ")");
            }
            if (this.device.getNews().size() > 0) {
                fragments.add(new PostsFragment().setSource(PostsFragment.SRC_NEWS).setDevice(this.device));
                titles.add(App.getInstance().getString(R.string.device_page_news) + " (" + this.device.getNews().size() + ")");
            }
            if (this.device.getFirmwares().size() > 0) {
                fragments.add(new PostsFragment().setSource(PostsFragment.SRC_FIRMWARES).setDevice(this.device));
                titles.add(App.getInstance().getString(R.string.device_page_firmwares) + " (" + this.device.getFirmwares().size() + ")");
            }
        }

        @Override
        public Fragment getItem(int position) {
            return fragments.get(position);
        }

        @Override
        public int getCount() {
            return fragments.size();
        }

        @Override
        public CharSequence getPageTitle(int position) {
            return titles.get(position);
        }
    }


    public class ImagesAdapter extends PagerAdapter {
        //private SparseArray<View> views = new SparseArray<>();
        private LayoutInflater inflater;
        private ArrayList<String> urls;
        ArrayList<String> fullUrls;

        public ImagesAdapter(Context context, ArrayList<String> urls, ArrayList<String> fullUrls) {
            this.inflater = LayoutInflater.from(context);
            this.urls = urls;
            this.fullUrls = fullUrls;
        }


        @Override
        public int getCount() {
            return urls.size();
        }

        @Override
        public Object instantiateItem(ViewGroup container, int position) {
            View imageLayout = inflater.inflate(R.layout.device_image_page, container, false);
            assert imageLayout != null;
            imageLayout.setOnClickListener(v -> ImageViewerActivity.startActivity(DeviceFragment.this.getContext(), fullUrls, position));
            container.addView(imageLayout, 0);
            loadImage(imageLayout, position);
            return imageLayout;
        }

        @Override
        public void destroyItem(ViewGroup container, int position, Object object) {
            container.removeView((View) object);
        }

        @Override
        public boolean isViewFromObject(View view, Object object) {
            return view.equals(object);
        }

        private void loadImage(View imageLayout, int position) {
            assert imageLayout != null;
            ImageView imageView = (ImageView) imageLayout.findViewById(R.id.image_view);
            ProgressBar progressBar = (ProgressBar) imageLayout.findViewById(R.id.progress_bar);
            ImageLoader.getInstance().displayImage(urls.get(position), imageView, new SimpleImageLoadingListener() {
                @Override
                public void onLoadingStarted(String imageUri, View view) {
                    progressBar.setVisibility(View.VISIBLE);
                }

                @Override
                public void onLoadingCancelled(String imageUri, View view) {
                    progressBar.setVisibility(View.GONE);
                }

                @Override
                public void onLoadingComplete(String imageUri, View view, Bitmap loadedImage) {
                    progressBar.setVisibility(View.GONE);
                }

                @Override
                public void onLoadingFailed(String imageUri, View view, FailReason failReason) {
                    progressBar.setVisibility(View.GONE);
                }
            });

        }
    }

}

File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/devdb/device/DeviceFragment.kt
package forpdateam.ru.forpda.ui.fragments.devdb.device

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Color
import android.os.Bundle
import android.view.Gravity
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.core.view.doOnLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.lifecycleScope
import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.appbar.CollapsingToolbarLayout
import com.google.android.material.tabs.TabLayout
import com.nostra13.universalimageloader.core.ImageLoader
import com.nostra13.universalimageloader.core.assist.FailReason
import com.nostra13.universalimageloader.core.listener.SimpleImageLoadingListener
import com.robohorse.pagerbullet.PagerBullet
import forpdateam.ru.forpda.R
import forpdateam.ru.forpda.databinding.FragmentDeviceBinding
import forpdateam.ru.forpda.databinding.ToolbarDeviceBinding
import forpdateam.ru.forpda.entity.remote.devdb.Device
import forpdateam.ru.forpda.extensions.getColorFromAttr
import forpdateam.ru.forpda.extensions.getDimenPx
import forpdateam.ru.forpda.extensions.quillMoxyPresenter
import forpdateam.ru.forpda.extensions.setBackgroundAttr
import forpdateam.ru.forpda.extensions.setBackgroundTintColor
import forpdateam.ru.forpda.presentation.devdb.device.DevicePresenter
import forpdateam.ru.forpda.presentation.devdb.device.DeviceView
import forpdateam.ru.forpda.ui.DimensionHelper
import forpdateam.ru.forpda.ui.DimensionsProvider
import forpdateam.ru.forpda.ui.activities.imageviewer.ImageViewerActivity
import forpdateam.ru.forpda.ui.fragments.TabFragment
import forpdateam.ru.forpda.ui.fragments.devdb.DevDbHelper
import forpdateam.ru.forpda.ui.fragments.devdb.device.comments.CommentsFragment
import forpdateam.ru.forpda.ui.fragments.devdb.device.posts.PostsFragment
import forpdateam.ru.forpda.ui.fragments.devdb.device.specs.SpecsFragment
import forpdateam.ru.forpda.ui.fragments.notes.NotesAddPopup
import forpdateam.ru.forpda.ui.fragments.tabBinding
import forpdateam.ru.forpda.ui.fragments.tabToolbarBinding
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import ru.radiationx.quill.inject

/**
 * Created by radiationx on 08.08.17.
 */

class DeviceFragment : TabFragment(R.layout.fragment_device), DeviceView {

    private val binding by tabBinding(FragmentDeviceBinding::bind)
    private val toolbarBinding by tabToolbarBinding(ToolbarDeviceBinding::bind)

    private val imagesPager: PagerBullet
        get() = toolbarBinding.imagesPager
    private val rating: TextView
        get() = toolbarBinding.itemRating
    private val fragmentsPager: ViewPager
        get() = binding.viewPager
    private val progressBar: ProgressBar
        get() = binding.progressBar
    private val toolbarContent: RelativeLayout
        get() = toolbarBinding.root

    private val dimensionsProvider by inject<DimensionsProvider>()

    private lateinit var copyLinkMenuItem: MenuItem
    private lateinit var shareMenuItem: MenuItem
    private lateinit var noteMenuItem: MenuItem
    private lateinit var toBrandMenuItem: MenuItem
    private lateinit var toBrandsMenuItem: MenuItem

    private var appBarOffset = 0

    private val presenter by quillMoxyPresenter<DevicePresenter>()

    init {
        configuration.defaultTitle = getString(R.string.fragment_title_device)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.apply {
            presenter.deviceId = getString(ARG_DEVICE_ID, null)
        }

        val transaction = childFragmentManager.beginTransaction()
        for (fragment in childFragmentManager.fragments) {
            transaction.remove(fragment)
        }
        transaction.commit()
        childFragmentManager.executePendingTransactions()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        baseInflateToolbar(R.layout.toolbar_device)

        val tabLayout = TabLayout(requireContext())
        val tabParams = CollapsingToolbarLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT,
            Gravity.BOTTOM
        )
        tabParams.collapseMode = CollapsingToolbarLayout.LayoutParams.COLLAPSE_MODE_PIN
        tabLayout.layoutParams = tabParams
        toolbarLayout.addView(tabLayout)

        val params = toolbarLayout.layoutParams as AppBarLayout.LayoutParams
        params.scrollFlags =
            AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL or AppBarLayout.LayoutParams.SCROLL_FLAG_EXIT_UNTIL_COLLAPSED
        toolbarLayout.layoutParams = params

        val newParams = toolbar.layoutParams as CollapsingToolbarLayout.LayoutParams
        newParams.collapseMode = CollapsingToolbarLayout.LayoutParams.COLLAPSE_MODE_PIN
        newParams.bottomMargin = toolbar.context.getDimenPx(R.dimen.dp48)
        toolbar.layoutParams = newParams
        toolbar.requestLayout()

        setCardsBackground()
        toolbarTitleView.setShadowLayer(
            toolbarTitleView.context.getDimenPx(R.dimen.dp2).toFloat(),
            0f,
            0f,
            toolbarTitleView.context.getColorFromAttr(androidx.appcompat.R.attr.colorPrimary)
        )
        toolbarSubtitleView.setShadowLayer(
            toolbarSubtitleView.context.getDimenPx(R.dimen.dp2).toFloat(),
            0f,
            0f,
            toolbarSubtitleView.context.getColorFromAttr(androidx.appcompat.R.attr.colorPrimary)
        )

        toolbarLayout.setExpandedTitleColor(Color.TRANSPARENT)
        toolbarLayout.setCollapsedTitleTextColor(Color.TRANSPARENT)
        toolbarLayout.isTitleEnabled = false

        tabLayout.tabMode = TabLayout.MODE_SCROLLABLE
        tabLayout.setupWithViewPager(fragmentsPager)

        imagesPager.setIndicatorTintColorScheme(
            imagesPager.context.getColorFromAttr(R.attr.default_text_color),
            imagesPager.context.getColorFromAttr(R.attr.second_text_color)
        )

        appBarLayout.addOnOffsetChangedListener(AppBarLayout.OnOffsetChangedListener { _, offset ->
            appBarOffset = offset
            updateToolbarShadow()
        })

        if (configuration.isFitSystemWindow) {
            dimensionsProvider
                .observeDimensions()
                .onEach { dimensions ->
                    toolbarContent.doOnLayout {
                        updateDimens(dimensions)
                    }
                    updateDimens(dimensions)
                }
                .launchIn(viewLifecycleOwner.lifecycleScope)
        }
    }

    override fun isShadowVisible(): Boolean {
        return appBarOffset != 0
    }

    private fun updateDimens(dimensions: DimensionHelper.Dimensions) {
        toolbarContent?.also {
            val params = it.layoutParams as CollapsingToolbarLayout.LayoutParams
            params.topMargin = dimensions.statusBar
            it.layoutParams = params
        }
    }

    override fun addBaseToolbarMenu(menu: Menu) {
        super.addBaseToolbarMenu(menu)
        copyLinkMenuItem = menu.add(R.string.copy_link)
            .setOnMenuItemClickListener {
                presenter.copyLink()
                true
            }

        shareMenuItem = menu.add(R.string.share)
            .setOnMenuItemClickListener {
                presenter.shareLink()
                true
            }

        noteMenuItem = menu.add(R.string.create_note)
            .setOnMenuItemClickListener {
                presenter.createNote()
                true
            }

        toBrandMenuItem = menu.add(R.string.devices)
            .setOnMenuItemClickListener {
                presenter.openDevices()
                true
            }

        toBrandsMenuItem = menu.add(R.string.devices)
            .setOnMenuItemClickListener {
                presenter.openBrands()
                true
            }

        refreshToolbarMenuItems(false)
    }

    override fun refreshToolbarMenuItems(enable: Boolean) {
        super.refreshToolbarMenuItems(enable)
        if (enable) {
            copyLinkMenuItem.isEnabled = true
            shareMenuItem.isEnabled = true
            noteMenuItem.isEnabled = true
            toBrandMenuItem.isVisible = true
            toBrandsMenuItem.isVisible = true
        } else {
            copyLinkMenuItem.isEnabled = false
            shareMenuItem.isEnabled = false
            noteMenuItem.isEnabled = false
            toBrandMenuItem.isVisible = false
            toBrandsMenuItem.isVisible = false
        }
    }

    override fun showData(data: Device) {
        progressBar.visibility = View.GONE
        toBrandMenuItem.title = "${data.catTitle} ${data.brandTitle}"
        toBrandsMenuItem.title = data.catTitle
        refreshToolbarMenuItems(true)
        setTitle(data.title)
        setTabTitle("${data.catTitle} ${data.brandTitle}: ${data.title}")
        setSubtitle("${data.catTitle} ${data.brandTitle}")


        val urls = ArrayList<String>()
        val fullUrls = ArrayList<String>()
        for (pair in data.images) {
            urls.add(pair.first)
            fullUrls.add(pair.second)
        }
        val imagesAdapter = ImagesAdapter(requireContext(), urls, fullUrls)
        imagesPager.setAdapter(imagesAdapter)

        val pagerAdapter = FragmentPagerAdapter(childFragmentManager, data)
        fragmentsPager.adapter = pagerAdapter

        if (data.rating > 0) {
            rating.text = data.rating.toString()
            rating.setBackgroundAttr(R.attr.count_background)
            rating.setBackgroundTintColor(DevDbHelper.getColor(data.rating))
            rating.visibility = View.VISIBLE
            if (!data.comments.isEmpty()) {
                rating.isClickable = true
                rating.setOnClickListener { fragmentsPager.setCurrentItem(1, true) }
            }

        } else {
            rating.visibility = View.GONE
        }
    }

    override fun showCreateNote(title: String, url: String) {
        NotesAddPopup.showAddNoteDialog(requireContext(), title, url)
    }

    private inner class FragmentPagerAdapter(
        fm: FragmentManager,
        private val device: Device
    ) : androidx.fragment.app.FragmentPagerAdapter(fm) {
        private val fragments = ArrayList<Fragment>()
        private val titles = ArrayList<String>()

        init {
            if (!this.device.specs.isEmpty()) {
                fragments.add(SpecsFragment().setDevice(this.device))
                titles.add(getString(R.string.device_page_specs))
            }
            if (!this.device.comments.isEmpty()) {
                fragments.add(CommentsFragment().setDevice(this.device))
                val title = getString(R.string.device_page_comments, this.device.comments.size)
                titles.add(title)
            }
            if (!this.device.discussions.isEmpty()) {
                fragments.add(
                    PostsFragment().setSource(PostsFragment.SRC_DISCUSSIONS).setDevice(this.device)
                )
                val title = getString(R.string.device_page_discussions, this.device.discussions.size)
                titles.add(title)
            }
            if (!this.device.news.isEmpty()) {
                fragments.add(
                    PostsFragment().setSource(PostsFragment.SRC_NEWS).setDevice(this.device)
                )
                val title = getString(R.string.device_page_news, this.device.news.size)
                titles.add(title)
            }
            if (!this.device.firmwares.isEmpty()) {
                fragments.add(
                    PostsFragment().setSource(PostsFragment.SRC_FIRMWARES).setDevice(this.device)
                )
                val title = getString(R.string.device_page_firmwares, this.device.firmwares.size)
                titles.add(title)
            }
        }

        override fun getItem(position: Int): Fragment {
            return fragments[position]
        }

        override fun getCount(): Int {
            return fragments.size
        }

        override fun getPageTitle(position: Int): CharSequence {
            return titles[position]
        }
    }


    inner class ImagesAdapter(
        context: Context,
        private val urls: ArrayList<String>,
        private var fullUrls: ArrayList<String>
    ) : PagerAdapter() {
        //private SparseArray<View> views = new SparseArray<>();
        private val inflater: LayoutInflater = LayoutInflater.from(context)


        override fun getCount(): Int {
            return urls.size
        }

        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            val imageLayout = inflater.inflate(R.layout.device_image_page, container, false)
            imageLayout.setOnClickListener {
                ImageViewerActivity.startActivity(
                    this@DeviceFragment.requireContext(),
                    fullUrls,
                    position
                )
            }
            container.addView(imageLayout, 0)
            loadImage(imageLayout, position)
            return imageLayout
        }

        override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
            container.removeView(`object` as View)
        }

        override fun isViewFromObject(view: View, `object`: Any): Boolean {
            return view == `object`
        }

        private fun loadImage(imageLayout: View, position: Int) {
            val imageView = imageLayout.findViewById<View>(R.id.image_view) as ImageView
            val progressBar = imageLayout.findViewById<View>(R.id.progress_bar) as ProgressBar
            ImageLoader.getInstance()
                .displayImage(urls[position], imageView, object : SimpleImageLoadingListener() {
                    override fun onLoadingStarted(imageUri: String?, view: View?) {
                        progressBar.visibility = View.VISIBLE
                    }

                    override fun onLoadingCancelled(imageUri: String?, view: View?) {
                        progressBar.visibility = View.GONE
                    }

                    override fun onLoadingComplete(
                        imageUri: String?,
                        view: View?,
                        loadedImage: Bitmap?
                    ) {
                        progressBar.visibility = View.GONE
                    }

                    override fun onLoadingFailed(
                        imageUri: String?,
                        view: View?,
                        failReason: FailReason?
                    ) {
                        progressBar.visibility = View.GONE
                    }
                })

        }
    }

    companion object {
        const val ARG_DEVICE_ID = "DEVICE_ID"
    }

}
